### ISSUE TITLE
Nearest Neighbor lookup doesn't find date-time when it should exist

### ISSUE THREAD

**Issue by mpu-creare** [creare-com/podpac#155]
**Description**
When I evaluate a SMAP node with a date that should exist using 'nearest' interpolator, the time is not found. 

**Steps to Reproduce**
```python
>>> import podpac
>>> smap = podpac.datalib.smap.SMAP(product='SPL3SMP')
>>> nc = smap.find_coordinates()[0]
>>> nc['time']
ArrayCoordinates1d(time): Bounds[2015-03-31, 2018-06-03], N[1157], ctype['midpoint']
>>> c = podpac.coordinates.merge_dims([nc.drop('time'), podpac.Coordinates(['2018-05-19T12'], ['time'])])
>>> smap.eval(c)  # <-- returns all nans
```

The specific problem seem to be this call:
```python
>>> smap.source_coordinates.intersect(c)
Coordinates
	time: ArrayCoordinates1d(time): Bounds[nan, nan], N[0], ctype['midpoint']
```

This works just fine:
```python
>>> c = podpac.coordinates.merge_dims([nc.drop('time'), podpac.Coordinates([['2018-05-19', '2018-05-20']], ['time'])])
>>> o = smap.eval(c)  # <-- returns all nans
>>> o.mean()
<xarray.UnitsDataArray ()>
array(0.219963)
```

**Comment by mpu-creare** [comment]
Incidentally, I've just realized that the compositors may need additional logic to, for example, interpolation between different files in the compositor...

**Comment by mlshapiro** [comment]
updating example to have username/password

```python
import podpac

smap = podpac.datalib.smap.SMAP(product='SPL3SMP', username=username, password=password)
smap.source_coordinates
nc = smap.find_coordinates()[0]
c = podpac.coordinates.merge_dims([nc.drop('time'), podpac.Coordinates(['2018-05-19T12'], ['time'])])
out = smap.eval(c)  # <-- returns all nans
```

**Comment by mlshapiro** [comment]
@jmilloy  this looks like a coordinates intersect handling issue when the length of the intesercted coordinate is 1:

```
from podpac import Coordinates, clinspace

a = Coordinates([clinspace('2015-03-31', '2018-06-03', 1157)], ['time'])
b = Coordinates([clinspace(-83.63197326660156, 83.63196563720703, 406), 
                 clinspace(-179.8129119873047, 179.8129119873047, 964), 
                 '2018-05-19T12'], ['lat', 'lon','time'])

a.intersect(b)
```

since both of these are `ArrayCoordinates1d`, it goes to the `select` method, but `bounds[0] == bounds[1]`

I don't have time to fix today, but I will attempt to fix and have you review

**Comment by jmilloy** [comment]
This will be fixed in #120. It may also require #103.

**Comment by jmilloy** [comment]
I should say, the intersection issue @mlshapiro points out will be fixed.

**Comment by mlshapiro** [comment]
@jmilloy I misspoke  - the example cited above does return nan values, but that is appropriate behavior as I wrote it since the default for the `outer` keyward arg is false. If I set `outer=True`, then I get back the outer two data points as expected. 


```
from podpac import Coordinates, clinspace

a = Coordinates([clinspace('2015-03-31', '2018-06-03', 1157)], ['time'])
b = Coordinates([clinspace(-83.63197326660156, 83.63196563720703, 406), 
                 clinspace(-179.8129119873047, 179.8129119873047, 964), 
                 '2018-05-19T12'], ['lat', 'lon','time'])

a.intersect(b, outer=True)
>>> Coordinates
	time: UniformCoordinates1d(time): Bounds[2018-05-19, 2018-05-20], N[2], ctype['midpoint']
```

I will continue to work on fixing this issue as @mpu-creare  has described it initially

**Comment by mlshapiro** [comment]
The smap intersect example also works with the `outer` keyword:

```python
smap.source_coordinates.intersect(c, outer=True)
```

By default the DataSource eval method sets `outer=True` during the intersect, so this is not related to the intersect

**Comment by mlshapiro** [comment]
f27778fef85e3069321874ca80af1f3ba0945753 implements the default interpolator definition in compositor nodes, but this was also not the issue

**Comment by jmilloy** [comment]
There is a short-circuit in the select method that uses the area_bounds to detect an empty selection. The area_bounds are incorrect for non-monotonic coordinates (unless you specify `ctype='point'` or provide `extents`), see #162. The area_bounds could also be incorrect for singleton coordinates. Are the SMAP times singletons or non-monotonic somewhere, in a way that makes them different from the examples you've posted?

**Comment by mpu-creare** [comment]
I believe it's all monotonic, but the requested coordinates are singleton.

**Comment by jmilloy** [comment]
I just tested this again and it is still evaluating to NaNs after #162 and #120 and #163.

Additionally, I tried evaluating at `2018-05-19` instead of `2018-05-19T12` and got a different error. (Note that `['2018-05-19', '2018-05-20']` works just fine).

```
In [41]: c = podpac.coordinates.merge_dims([nc.drop('time'), podpac.Coordinates(['2018-05-19T12'], ['time'])])

In [42]: o = smap.eval(c)

In [43]: np.isfinite(o).sum()
Out[43]: 
<xarray.DataArray ()>
array(0)

In [44]: c = podpac.coordinates.merge_dims([nc.drop('time'), podpac.Coordinates(['2018-05-19'], ['time'])])

In [45]: o = smap.eval(c)
/home/jmilloy/Pipeline/podpac/podpac/core/coordinates/coordinates1d.py:114: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  if not np.issubdtype(val.dtype, self.deltatype):
INFO:pydap:Fetching URL: https://n5eil02u.ecs.nsidc.org/opendap/SMAP/SPL3SMP.004/2018.05.19/SMAP_L3_SM_P_20180519_R15182_001.h5.dods?Soil_Moisture_Retrieval_Data_AM_soil_moisture[0:1:405][0:1:963]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-45-1d2eb0579d3b> in <module>()
----> 1 o = smap.eval(c)

~/Pipeline/podpac/podpac/core/node.py in wrapper(self, coordinates, output)
    561             self._from_cache = True
    562         else:
--> 563             data = fn(self, coordinates, output=output,)
    564             if self.cache_output:
    565                 self.put_cache(data, key, cache_coordinates, overwrite=self.cache_update,

~/Pipeline/podpac/podpac/core/compositor.py in eval(self, coordinates, output)
    251 
    252         outputs = self.iteroutputs(coordinates)
--> 253         output = self.composite(outputs, output)
    254         return output
    255 

~/Pipeline/podpac/podpac/core/compositor.py in composite(self, outputs, result)
    302         if result is None:
    303             # consume the first source output
--> 304             result = next(outputs).copy()
    305 
    306         # initialize the mask

~/Pipeline/podpac/podpac/core/compositor.py in iteroutputs(self, coordinates)
    227             output = None # scratch space
    228             for src in src_subset:
--> 229                 output = src.eval(coordinates, output)
    230                 yield output
    231                 #output[:] = np.nan

~/Pipeline/podpac/podpac/core/node.py in wrapper(self, coordinates, output)
    561             self._from_cache = True
    562         else:
--> 563             data = fn(self, coordinates, output=output,)
    564             if self.cache_output:
    565                 self.put_cache(data, key, cache_coordinates, overwrite=self.cache_update,

~/Pipeline/podpac/podpac/core/compositor.py in eval(self, coordinates, output)
    251 
    252         outputs = self.iteroutputs(coordinates)
--> 253         output = self.composite(outputs, output)
    254         return output
    255 

~/Pipeline/podpac/podpac/core/compositor.py in composite(self, outputs, result)
    302         if result is None:
    303             # consume the first source output
--> 304             result = next(outputs).copy()
    305 
    306         # initialize the mask

~/Pipeline/podpac/podpac/core/compositor.py in iteroutputs(self, coordinates)
    227             output = None # scratch space
    228             for src in src_subset:
--> 229                 output = src.eval(coordinates, output)
    230                 yield output
    231                 #output[:] = np.nan

~/Pipeline/podpac/podpac/core/data/datasource.py in eval(self, coordinates, output)
    299         # TODO move this if WCS can be updated to support
    300         self._evaluated_coordinates = coordinates
--> 301         return self._eval(coordinates, output=output)
    302 
    303     @node_eval

~/Pipeline/podpac/podpac/core/node.py in wrapper(self, coordinates, output)
    561             self._from_cache = True
    562         else:
--> 563             data = fn(self, coordinates, output=output,)
    564             if self.cache_output:
    565                 self.put_cache(data, key, cache_coordinates, overwrite=self.cache_update,

~/Pipeline/podpac/podpac/core/data/datasource.py in _eval(self, coordinates, output)
    342                                                  self._requested_source_data,
    343                                                  coordinates,
--> 344                                                  output)
    345 
    346         # return the output to the originally requested output dims

~/Pipeline/podpac/podpac/core/data/interpolate.py in interpolate(self, source_coordinates, source_data, eval_coordinates, output_data)
   1210 
   1211             if eq:
-> 1212                 output_data.data = source_data.data
   1213                 return output_data
   1214 

~/Pipeline/env3/lib/python3.5/site-packages/xarray-0.10.3-py3.5.egg/xarray/core/common.py in __setattr__(self, name, value)
    188                     "style assignment (e.g., `ds['name'] = ...`) instead to "
    189                     "assign variables." % (name, type(self).__name__))
--> 190         object.__setattr__(self, name, value)
    191 
    192     def __dir__(self):

~/Pipeline/env3/lib/python3.5/site-packages/xarray-0.10.3-py3.5.egg/xarray/core/dataarray.py in data(self, value)
    397     @data.setter
    398     def data(self, value):
--> 399         self.variable.data = value
    400 
    401     @property

~/Pipeline/env3/lib/python3.5/site-packages/xarray-0.10.3-py3.5.egg/xarray/core/variable.py in data(self, data)
    300         if data.shape != self.shape:
    301             raise ValueError(
--> 302                 "replacement data must match the Variable's shape")
    303         self._data = data
    304 

ValueError: replacement data must match the Variable's shape
```

**Comment by mpu-creare** [comment]
I think the SMAP compositor still needs some update to incorporate your changes... also potentially we really need the compositor to do some interpolation for us.

**Comment by jmilloy** [comment]
Ah, for reference that's #195.

**Comment by mpu-creare** [comment]
closing in favor of #195
